metaflow/runtime.py

```diff
@@ -238,8 +245,32 @@ def execute(self):
                                         'by the end of flow.')
 
     def _killall(self):
+        # If we are here, all children have received a signal and are shutting down.
+        # We want to give them an opportunity to do so and then kill
+        live_workers = []
+        temp_workers = {}
         for worker in self._workers.values():
-            worker.kill()
+            temp_workers[worker] = 1
+        live_workers = temp_workers.keys()
+        now = int(time.time())
+        self._logger('Waiting for %d workers to shutdown...' % len(live_workers),
+                     system_msg=True, bad=True)
+        while live_workers and int(time.time()) - now < 5:
+            # While not all workers are dead and we have waited less than 5 seconds
+            new_live_workers = []
+            for worker in live_workers:
+                if not worker.pre_kill():
+                    new_live_workers.append(worker)
+            live_workers = new_live_workers
```

Comment:
minor syntactic nit (up to you to incorporate it) - `live_workers = [worker for worker in live_workers if not worker.pre_kill()]`

Reply:
Also, is the intended behavior that every worker gets the `pre_kill` signal at least once, even if it takes more than 5 seconds?

Reply:
Yes, I thought of checking every time in the loop but decided against it because it would be less clean and nothing in the loop is blocking (they are all polls) so it should be relatively fast in all cases so it may take slightly more than 5 seconds but I doubt it will go longer.